### ISSUE TITLE
Question: Can multiple challenges be made in parallel and then satisfied all at once?

### ISSUE THREAD

**Issue by gitgrimbo** [Daplie/node-letsencrypt#59]
Hi,

My scenario is for a certificate with multiple domains.  It could be that I'm misunderstanding the ACME spec.

- I am requesting a certificate from a machine that isn't the web server that will use the certificate.
- I want, say, 5 domains to be associated with the certificate.
- For each domain I have to satisfy a challenge.
- In my scenario this means I have to do 5 deployments to my web server so that it can satisfy these challenges, one by one.

Is there a way I can set the 5 challenges in motion and do a single deployment to the web server that contains all the challenge responses?

Thanks.

**Comment by coolaj86** [comment]
I think that it's being serialized in `le-acme-core` and I think it could be changed to run parallel without breaking the API, but it's not something I want to work on right now.

If it's something you want to work on I'd just like to have an option for how many to run in parallel at a time with a reasonable default - like 4.

**Comment by coolaj86** [comment]
Each domain will be contacted by the ACME server. That's part of the spec.

**Comment by gitgrimbo** [comment]
Yeah, for my purposes it's not so much the number of ACME server requests made to my domains that's an issue, more the fact that I'd like to prepare my domain(s) all at once for the ACME requests (as all my subdomains are actually handled by the same web server, and thus would benefit from a single update/deployment).

**Comment by insightfuls** [comment]
How do you intend to deploy? Doing a manual deployment every 80 days I suspect would quickly become tiresome, so automating or at least semi-automating deployment is highly recommended (it's how ACME/Let's Encrypt was intended to be used). If it's automated, it shouldn't really be a problem doing multiple deployments, should it? I'd rather work on something that makes your deployment easier, in line with the ACME philosophy, than work on changing the code to facilitate going against the ACME philosophy, if possible.

**Comment by gitgrimbo** [comment]
My particular scenario is a Google App Engine Java project. Currently I have to do a deployment to get the challenge response on the server. One deployment per domain name associated with the certificate request, which in my case is 4 domains. I'd like to get this down to a single deployment, but I have already done 4 deployments for my certificate in the past and it was fine, if a little slow.

I could do something more exotic and host/deploy the challenge responses on a different server (e.g. a private GitHub repo or something), and have my App Engine project read and replay the challenges from there, but I'm looking at challenge batches first.

Until this issue is resolved, https://code.google.com/p/googleappengine/issues/detail?id=12535, I have to do manual certificate installation anyway.

**Comment by insightfuls** [comment]
I see.

Well, as @coolaj86 pointed out, it's the logic in `le-acme-core` that controls the order, specifically in `get-certificate.js`; the getCert function kicks everything off by calling `nextDomain` which orchestrates the whole thing; `challengeDone` calls `ensureValidation` after `options.setChallenge` has run its course (to deploy the challenge response for a domain). If you took that out and modified `nextDomain` to do a second phase for `ensureValidation` after the `getChallenges` phase had done all domains, you should be able to get the behaviour you want. You do have to make sure everything in the `state` is correctly maintained, though (e.g. the domain currently being validated is set there in `getChallenges`); I'm not sure how much modification it might need to handle challenges for multiple domains concurrently.

**Comment by gitgrimbo** [comment]
Yeah thanks for the tips.  I've 'promisified' `get-certificate.js` to get me started, but haven't done much more than that yet.

I've started to look at deploying a lightweight webapp to store the letsencrypt challenges, and have my main webapp proxy letsencrypt challenges to there.

**Comment by coolaj86** [comment]
FYI: Let's Encrypt v2 now supports wildcard domains.

Supported landed in greenlock.js v2.2.4 at https://git.coolaj86.com/coolaj86/greenlock.js
